### ISSUE TITLE
etcd3 and kube-apiserver fail on terraform apply after terraform destroying w/ kops generated config

### ISSUE THREAD

**Issue by austinorth** [kubernetes/kops#7414]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.** 
`Version 1.12.3 (git-e55205471)`
**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.10", GitCommit:"e3c134023df5dea457638b614ee17ef234dc34a6", GitTreeState:"clean", BuildDate:"2019-07-08T03:50:59Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
Unable to connect to the server: EOF
```
**3. What cloud provider are you using?**
AWS
**4. What commands did you run?  What is the simplest way to reproduce this issue?**
- `kops edit cluster` to configure.
- `kops update cluster --out=./ --target=terraform` to generate the Terraform config
- `terraform apply` to apply the config.

**5. What happened after the commands executed?**
- The `terraform apply` completes successfully and all instances are created. However, I cannot connect to the cluster even after 30 minutes and all nodes are marked as OutOfService by the load balancer. 
- Running a `kops validate cluster` results in: 
```unexpected error during validation: error listing nodes: Get https://api.example.net/api/v1/nodes: EOF```
- `kubectl get cluster` results in:
```Unable to connect to the server: EOF```
- `ssh` into the master nodes and I find
```
0815 14:23:15.697247       1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry [https://127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt true true 1000 0xc420b3ddd0 <nil> 5m0s 1m0s}), err (dial tcp 127.0.0.1:4001: connect: connection refused)
```
**6. What did you expect to happen?**
I expected the cluster to be up and ready to go, and to be able to connect to it.
**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**
```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 2019-07-31T15:59:08Z
  name: example.net
spec:
  additionalPolicies:
    node: |
      [   
        {
          "Effect": "Allow",
          "Action": [
            "sts:AssumeRole"
          ],
          "Resource": [
            "arn:aws:iam:::role/k8s-*"
          ]
        }
      ]   
  api:
    loadBalancer:
      sslCertificate: arn:aws:acm:us-east-1:secretsecretsecret
      type: Internal
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws 
  configBase: s3://kops.example.internal/example.net
  dnsZone: example.net
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-east-1b
      name: b
    - instanceGroup: master-us-east-1d
      name: d
    - instanceGroup: master-us-east-1e
      name: e
    name: main
  - etcdMembers:
    - instanceGroup: master-us-east-1b
      name: b
    - instanceGroup: master-us-east-1d
      name: d
    - instanceGroup: master-us-east-1e
      name: e
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
  kubernetesApiAccess:
  - office_cidr
  - another_cidr
  kubernetesVersion: 1.12.10
  masterInternalName: api.internal.example.net
  masterPublicName: api.example.net
  networkCIDR: another_cidr
  networkID: vpc-abcdefg
  networking:
    weave:
      mtu: 8912
  nonMasqueradeCIDR: another_cidr
  sshAccess:
  - another_cidr
  - another_cidr
  subnets:
  - cidr: another_cidr
    egress: nat-0e0e0e0e0e0e0
    id: subnet-abcdefg
    name: us-east-1b
    type: Private
    zone: us-east-1b
  - cidr: another_cidr
    egress: nat-0e0e0e0e0e0e0
    id: subnet-abcdefg
    name: us-east-1d
    type: Private
    zone: us-east-1d
  - cidr: another_cidr
    egress: nat-0e0e0e0e0e0e0
    id: subnet-abcdefg
    name: us-east-1e
    type: Private
    zone: us-east-1e
  - cidr: another_cidr
    id: subnet-abcdefg
    name: utility-us-east-1b
    type: Utility
    zone: us-east-1b
  - cidr: another_cidr
    id: subnet-abcdefg
    name: utility-us-east-1d
    type: Utility
    zone: us-east-1d
  - cidr: another_cidr
    id: subnet-abcdefg
    name: utility-us-east-1e
    type: Utility
    zone: us-east-1e
  topology:
    dns:
      type: Public
    masters: private
    nodes: private

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2019-08-02T02:43:49Z
  labels:
    kops.k8s.io/cluster: example.net
  name: master-us-east-1b
spec:
  image: kope.io/k8s-1.12-debian-stretch-amd64-hvm-ebs-2019-06-21
  machineType: m3.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1b
  role: Master
  subnets:
  - us-east-1b

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2019-08-02T02:43:49Z
  labels:
    kops.k8s.io/cluster: example.net
  name: master-us-east-1d
spec:
  image: kope.io/k8s-1.12-debian-stretch-amd64-hvm-ebs-2019-06-21
  machineType: m3.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1d
  role: Master
  subnets:
  - us-east-1d

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2019-08-02T02:43:50Z
  labels:
    kops.k8s.io/cluster: example.net
  name: master-us-east-1e
spec:
  image: kope.io/k8s-1.12-debian-stretch-amd64-hvm-ebs-2019-06-21
  machineType: m3.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1e
  role: Master
  subnets:
  - us-east-1e

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2019-08-02T02:43:50Z
  labels:
    kops.k8s.io/cluster: example.net
  name: nodes
spec:
  image: kope.io/k8s-1.12-debian-stretch-amd64-hvm-ebs-2019-06-21
  machineType: m4.xlarge
  maxSize: 3
  minSize: 3
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - us-east-1b
  - us-east-1d
  - us-east-1e


```
**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**
I don't think the verbose logs from the `kops` commands or the `terraform` commands are relevant here, and that the more relevant logs are those that I have posted below.
**9. Anything else do we need to know?**
- All of this works perfectly with version 1.11 of k8s and kops. We're moving forward with development on that version for now, but concerned about the future.
- If I `ssh` into the master nodes, the last message in `kube-apiserver.log` is 
```
0815 14:23:15.697247       1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry [https://127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt true true 1000 0xc420b3ddd0 <nil> 5m0s 1m0s}), err (dial tcp 127.0.0.1:4001: connect: connection refused)
```
- All three masters show this same message.
- Although it looks like the volume eventually attaches at a different mount point, an error that might be relevant in `/var/log/etcd.log` is
```
W0815 13:46:56.016362    3495 mounter.go:293] Error attaching volume "vol-0e7dd951edc73df76": Error attaching EBS volume "vol-0e7dd951edc73df76": InvalidParameterValue: Invalid value '/dev/xvdu' for unixDevice. Attachment point /dev/xvdu is already in use
        status code: 400, request id: 1cbf8ec1-49e1-4be0-8f20-487d79228007
```
- The cluster is running etcd version 3.2.24.

**Comment by mccare** [comment]
Your message looks like etcd-manager cannot mount its persistent EBS volume where etcd data lives (since it says 'already in use'). If you setup the cluster from scratch make sure you have no EBS volumes from your previous cluster (e.g. named a.etcd-main.clustername or tagged with your cluster name), or even duplicate EBS volumes with the same name.

In terraform you should see 2 EBS volumes per master created (aws_ebs_volume.a-etcd-events-...). I had trouble in the area of EBS volumes being created by terraform and an upgrade to kops 1.13 helped me.

**Comment by austinorth** [comment]
@mccare thanks for reaching out!
I verified we do not have any volumes from the previous cluster in EBS, and that Terraform plan output includes 2 EBS volumes per master. **Still experiencing the issue with kops 1.13.** Here is some example output from a Terraform plan, showing the two EBS volumes to be created for one of the masters:
```
  + aws_ebs_volume.d-etcd-events-example-net                                                                                                                                                                        
      id:                                             <computed>                                                                                                                                                
      arn:                                            <computed>                                                                                                                                                    
      availability_zone:                              "us-east-1d"                                                                                                                                              
      encrypted:                                      "false"                                                                                                                                                       
      iops:                                           <computed>                                                                                                                                                    
      kms_key_id:                                     <computed>                                                                                                                                                    
      size:                                           "20"                                                                                                                                                          
      snapshot_id:                                    <computed>                                                                                                                                                    
      tags.%:                                         "5"                                                                                                                                                           
      tags.KubernetesCluster:                         "example.net"                                                                                                                                                 
      tags.Name:                                      "d.etcd-events.example.net"                                                                                                                                  
      tags.k8s.io/etcd/events:                        "d/b,d,e"                                                                                                                                                     
      tags.k8s.io/role/master:                        "1"                                                                 
      tags.kubernetes.io/cluster/kstr8ip.net:         "owned"                                     
      type:                                           "gp2"                                                                       
                                                                                
  + aws_ebs_volume.d-etcd-main-example-net                                
      id:                                             <computed>                                         
      arn:                                            <computed>                                                                                                                                            
      availability_zone:                              "us-east-1d"                                                                                                                                                  
      encrypted:                                      "false"                                                                                                                                                       
      iops:                                           <computed>                                                                                                                                                    
      kms_key_id:                                     <computed>                                                                                                                                                    
      size:                                           "20"                                                                                                                                                          
      snapshot_id:                                    <computed>                                                                                                                                                    
      tags.%:                                         "5"                                                                                                                                                   
      tags.KubernetesCluster:                         "example.net"                                                               
      tags.Name:                                      "d.etcd-main.example.net"                     
      tags.k8s.io/etcd/main:                          "d/b,d,e"                               
      tags.k8s.io/role/master:                        "1"                                  
      tags.kubernetes.io/cluster/kstr8ip.net:         "owned"
      type:                                           "gp2"
```

**Comment by mccare** [comment]
Not an expert here, but from my logs I can tell:
- `Unable to create storage backend` is also in my kube-apiserver.log
- Main change between 1.11 and 1.12/1.13 is that now etcd-manager is used instead of directly etcd. Look into `/var/log/etcd.log` to see if etcd is reporting errors. 

The log starts out by looking for EBS volumes and then attaching them.

**Comment by alanmoment** [comment]
I think your problem is the same as mine.

My cluster stat:
kops 1.12.3
k8s upgrade from 1.10.x to 1.12.10
etcd upgrade from 3.0.17 to 3.2.24

I got error message "/opt/etcd-v3.0.17-linux-amd64 not found" in etcd.log. Maybe you too.

You can download file in https://github.com/etcd-io/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz and put in etcd-manage-main and etcd-manage-event container.

Then executing cluster rolling-update. 

Hope it's can help you.

**Comment by mamoit** [comment]
I just had this same issue, but only happened with one of the masters, so I didn't even notice until I did `kubectl get node` and only 2 showed up.
Running terraform `0.11.14` and kops `1.12.3`, this never happened to me with `1.12.2`.

**Comment by dandillingham** [comment]
I'm getting this at well, no terraform, just kops, using version 1.13.0.

**Comment by mamoit** [comment]
I force roll-updated the master nodes, and they came up fine.
```
kops rolling-update cluster yourclusternamehere --instance-group-roles=Master --force
```

**Comment by austinorth** [comment]
> I force roll-updated the master nodes, and they came up fine.
> 
> ```
> kops rolling-update cluster yourclusternamehere --instance-group-roles=Master --force
> ```

Ah! So apply everything with Terraform and then force roll achieves a working cluster?

**Comment by austinorth** [comment]
Still having issues with this. I followed these steps to get the cluster upgraded to 1.12.10 in our test account:
```
kops set cluster cluster.spec.etcdClusters[*].provider=manager
kops update cluster --out ./ --target=terraform
terraform apply
kops rolling-update cluster --yes

# Change to 1.12.10
kops edit cluster

# Update images to kope.io/k8s-1.12-debian-stretch-amd64-hvm-ebs-2019-08-16
kops edit instancegroups master-xxx
kops edit instancegroup nodes
kops update cluster --out ./ --target=terraform
terraform apply
kops rolling-update cluster --cloudonly --master-interval=1s --node-interval=1s --yes
```
And the upgrade itself runs fine, and the cluster is healthy when it comes back up. But then, when I fully spin _down_ the cluster (as we do in our test environment, since it doesn't need to be running all the time), then spin it back up, I get the following errors on the master nodes, which I believe to be the root of the problem.

- from `etcd-events.log`
```
W1008 13:54:13.464107    3399 controller.go:149] unexpected error running etcd cluster reconciliation loop: etcd has 0 members registered; must issue restore-backup command to proceed
```
which is in turn caused by...
- from `kube-apiserver.log`
```
F1008 13:52:35.532589       1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry [https://127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/k
ube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt true 0xc00098c3f0 <nil> 5m0s 1m0s}), err (dial tcp 127.0.0.1:4001: connect: connection refused) 
```
What upgrade steps am I missing here to allow the _new_ cluster to set up its storage backends properly? This doesn't seem to be documented anywhere.
For the record, I also ran @mamoit's command when I arrived at this state, which did not run because I could not reach the k8s api with the masters in this state. I added the `--cloudonly` flag to make it run despite that, and still no luck.

**Comment by austinorth** [comment]
As mentioned in #6605, I tried the fix [here](https://www.hindenes.com/2019-08-09-Kops-Restore/), and it brought back cluster functionality. However, it did _not_ alleviate the issue when the cluster is spun down using `terraform destroy` and spun back up using `terraform apply`. It seems to revert to the same broken state at this point. I will just try creating a whole new k8s 1.14 cluster next to see if there's any change.

**Comment by austinorth** [comment]
I was able to get Kubernetes version 1.14 to spin up successfully with the Terraform config generated by `kops create cluster`, but when I did a `terraform destroy` and a `terraform apply` to destroy and recreate the cluster, I experienced the same issue. This caused me to ponder why it would succeed on the first `terraform apply`, but not on any after the cluster had been destroyed.  

I reasoned there had to be some config stored _outside_ of what was created by Terraform that persisted after a `terraform destroy` that the _new_ cluster was picking up when initializing `etcd-manager`, which then caused it to error trying to reach a state that only would ever exist in the previous cluster. Looking in the s3 folder, my suspicions were quickly confirmed:
![image (3)](https://user-images.githubusercontent.com/31007285/66612588-3e076200-eb90-11e9-9945-6864cfd396d9.png)
These `etcd-cluster-created` files exist in two locations:
- `<s3-bucket>/<full-cluster-name>/backups/etcd/main/etcd-cluster-created`
- `<s3-bucket>/<full-cluster-name>/backups/etcd/events/control/etcd-cluster-created`

These two files cause the new cluster to look for the _old_ etcd EBS volumes, which were destroyed with the `terraform destroy`! Deleting them and _then_ running `terraform apply` to spin up the cluster resolves the issue. The fact that these files persist seems to actually be a feature though, in that it allows the cluster to recover, reattach to its original etcd3 volumes, and restore the original state easily if the masters go down.
